### PR TITLE
Feature/apache filters

### PR DESCRIPTION
**Pull request by fdurand**
For review please !
Added apache filters to be able to reject unusefull request like the weather widget or active sync from devices.

### CODE REVIEW

**Comment by fdurand** [comment]
Up !

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 1]
```diff
@@ -2850,26 +2850,33 @@ Avoid captive portal overload due to non-browser HTTP requests
...
 By default we allow every query to be redirected and reach PacketFence for the captive portal operation. In a lot of cases, this means that a lot of non-user initiated queries reach PacketFence and waste its resources for nothing since they are not from browsers. (iTunes, Windows update, MSN Messenger, Google Desktop, ...).
 
-So far, we blacklisted clients known to be misbehaving. However, a completely different approach can be taken: whitelist only known browsers.
+Since PacketFence 4.2 we introduced Apache filter directly in the configuration.
 
-This has the nasty side-effect of being unfriendly with (blocking) less popular browsers and devices so this is disabled by default.
+By default some rules has been enabled like refuse request from empty user-agent but you can have a look in the file apache_filters.conf to see samples.
 
-If you want to enable this feature, edit `conf/httpd.conf.d/block-unwanted.conf`, and uncomment the following lines:
+Filters are define with 2 blocks, the first one like [1] define a simple test, like:
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: defined

File: NEWS.asciidoc [hunk 2]
```diff
@@ -49,6 +49,7 @@ New Features
 * Integration of mod_qos in the Apache configuration of the captive portal
 * New pfcmd "cache" command
 * New pfcmd "configreload" command
+* Filters for HTTP request on the portal
```
Comment (cgx) [inline NEWS.asciidoc]: _HTTP requests_

File: conf/apache_filters.conf.example [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# Refuse undefined useragent
```
Comment (cgx) [inline conf/apache_filters.conf.example]: Instead of _Refuse_ I would use the term _Reject_.

File: conf/apache_filters.conf.example [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+redirect_url =
+
+# Refuse empty user agent
+[2]
+filter = user_agent
+method = POST
+operator = is
+regexp = ^$
+
+[2:2]
```
Comment (cgx) [inline conf/apache_filters.conf.example]: How about combining the two first rules (`[1] and [2]`) as `[1:1|2]`?

File: conf/apache_filters.conf.example [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Refuse undefined useragent
+# Define the test
+[1]
+filter = user_agent
+method = GET
+operator = is
+regexp = ^$
+# Define the logic and the action
```
Comment (cgx) [inline conf/apache_filters.conf.example]: Are the _tests/conditions_ and the _filters/actions_ required to be identified by integers? Could we use significant names instead? For example:
`[get_wo_user_agent]`
..
`[post_wo_user_agent]`
..
`[no_user_agent:get_wo_user_agent|post_wo_user_agent]`
..

File: conf/apache_filters.conf.example (hunk 3 above)
Comment (cgx) [inline conf/apache_filters.conf.example]: Be consistent and write _user agent_ (with a space) everywhere in the comments and documentation.

File: conf/apache_filters.conf.example [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+filter = uri
+method = GET
+operator = is_not
+regexp = /generate_204
+
+[4:4&5]
+action = 501
+redirect_url =
+
+## Only accept a list of useragent
```
Comment (cgx) [inline conf/apache_filters.conf.example]: .. _a list of user agents_

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 7]
```diff
@@ -2850,26 +2850,33 @@ Avoid captive portal overload due to non-browser HTTP requests
 
 By default we allow every query to be redirected and reach PacketFence for the captive portal operation. In a lot of cases, this means that a lot of non-user initiated queries reach PacketFence and waste its resources for nothing since they are not from browsers. (iTunes, Windows update, MSN Messenger, Google Desktop, ...).
 
-So far, we blacklisted clients known to be misbehaving. However, a completely different approach can be taken: whitelist only known browsers.
+Since PacketFence 4.2 we introduced Apache filter directly in the configuration.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: _Since version 4.2 of PacketFence, you can define HTTP filters for Apache from the configuration of PacketFence._

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 8]
```diff
@@ -2850,26 +2850,33 @@ Avoid captive portal overload due to non-browser HTTP requests
...
 
-This has the nasty side-effect of being unfriendly with (blocking) less popular browsers and devices so this is disabled by default.
+By default some rules has been enabled like refuse request from empty user-agent but you can have a look in the file apache_filters.conf to see samples.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: _Some rules have been enabled by default, like one to reject requests with no defined user agent. All rules, including some examples, are defined in the configuration file `apache_filters.conf`._

File: docs/PacketFence_Administration_Guide.asciidoc (hunk 1 above)
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: _Filters are defined with at least two blocks. First are the tests. For example:_

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 9]
```diff
@@ -2850,26 +2850,33 @@ Avoid captive portal overload due to non-browser HTTP requests
...
+  operator = is
+  regexp = Dalvik
+  
+  [2]
+  filter = uri
+  method = GET
+  operator = is_not
+  regexp = /generate_204
+  
+and the second one like [1:1&2] define the logic and the action, like:
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: _The last block defines the relationship between the tests and the desired action. For example:_

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 10]
```diff
@@ -2850,26 +2850,33 @@ Avoid captive portal overload due to non-browser HTTP requests
...
 
-This will allow the following browsers to reach the captive portal (but nothing else):
+  [1:1&2]
+  action = 501
+  redirect_url =
+
+A human readeable definition of this rule is:
+If user agent is Dalvik and uri is not /generate_204 then return the http code 501.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: _This filter will return an error code (501) if the user agent is Dalvik and the URI doesn't contain_ `_/generate_204`.

File: lib/pf/web/filter.pm [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+Test all the rules
+
+=cut
+
+sub test {
+    my ($self, $r) = @_;
+    my $logger = Log::Log4perl::get_logger( ref($self) );
+
+    foreach my $rule  ( sort keys %ConfigApacheFilters ) {
+        if ($rule =~ /^\d+:(.*)$/) {
```
Comment (cgx) [inline lib/pf/web/filter.pm]: As I mentioned, I think we could benefit from _not_ limiting the test id to be an integer.

File: lib/pf/web/filter.pm [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+                my $action = $self->dispatch_action($ConfigApacheFilters{$rule});
+                return ($action->($self,$r,$ConfigApacheFilters{$rule}));
+            }
+        }
+    }
+}
+
+=item dispatch_rules
+
+Return the reference to the function that parse the rule.
```
Comment (cgx) [inline lib/pf/web/filter.pm]: .._that parses the rule._

File: lib/pf/web/filter.pm [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    my $key = {
+        uri => \&uri_parser,
+        user_agent  => \&user_agent_parser,
+    };
+    return $key->{$rule->{'filter'}}->($self, $r, $rule);
+}
+
+=item dispatch_action
+
+Return the reference to the function that do the action.
```
Comment (cgx) [inline lib/pf/web/filter.pm]: .._that performs the action._

File: lib/pf/web/filter.pm [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+    if ($rule->{'action'} =~ /30\d/) {
+        return \&redirect;
+    } else {
+        return \&code;
+    }
+}
+
+=item uri_parser
+
+Parse the uri parameter and compare to the rule. If it match then take the action
```
Comment (cgx) [inline lib/pf/web/filter.pm]: _If it matches then perform the action._

File: lib/pf/web/filter.pm [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+
+=cut
+
+sub uri_parser {
+    my ($self, $r, $rule) = @_;
+
+    my $action;
+    if ($rule->{'operator'} eq 'is') {
+        if ((($r->uri =~ /$rule->{'regexp'}/) || ($r->args =~  /$rule->{'regexp'}/)) && ($r->method eq $rule->{'method'})) {
```
Comment (cgx) [inline lib/pf/web/filter.pm]: Instead of separately testing `uri` and `args`, I suggest to test on `unparsed_uri`.

File: lib/pf/web/filter.pm [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+            return 1;
+        } elsif (!$rule->{'action'}) {
+            return 0;
+        }
+    } else {
+        if ((($r->uri !~ /$rule->{'regexp'}/) && ($r->args !~  /$rule->{'regexp'}/)) && ($r->method eq $rule->{'method'})) {
```
Comment (cgx) [inline lib/pf/web/filter.pm]: Idem.

File: lib/pf/web/filter.pm [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+           return 1;
+        } elsif (!$rule->{'action'}) {
+            return 0;
+        }
+    }
+}
+
+=item user_agent_parser
+
+Parse user_agent parameter and compare to the rule. If it match take the action
```
Comment (cgx) [inline lib/pf/web/filter.pm]: _If it matches, perform the action._

File: lib/pf/web/filter.pm [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+    my ($self,$r,$rule) = @_;
+
+    my $action;
+    my $user_agent;
+    if (defined($r->headers_in->{'User-Agent'})) {
+        $user_agent = $r->headers_in->{'User-Agent'};
+    } else {
+        $user_agent = '';
+    }
+    if ($rule->{'operator'}eq 'is') {
```
Comment (cgx) [inline lib/pf/web/filter.pm]: Please put a space before the `eq` operator.

File: lib/pf/web/filter.pm [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+        $r->headers_out->set('Location' => $rule->{'redirect_url'});
+        return $rule->{'action'};
+    } else {
+        return 1;
+    }
+}
+
+=item code
+
+Return the apache code
```
Comment (cgx) [inline lib/pf/web/filter.pm]: _Return the HTTP code._

File: lib/pf/web/filter.pm [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+
+Minor parts of this file may have been contributed. See CREDITS.
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2014 Inverse inc.
+
+Copyright (C) 2005 Kevin Amorin
+
+Copyright (C) 2005 David LaPorte
```
Comment (cgx) [inline lib/pf/web/filter.pm]: Only keep the copyrights of Inverse.

File: conf/apache_filters.conf.example (hunk 5 above)
Comment (fdurand) [inline conf/apache_filters.conf.example]: Should be easy to replace [1:1&2] with [no_user_agent:1&2] but with [no_user_agent:get_wo_user_agent|post_wo_user_agent] will be a regexp nightmare.

**Pull request merged**